File: frontend/src/types/student.ts
export interface StudentInput {
  record_generated_date: string
  enrollment_date: string
  academic_year: string
  country: string
  state_province: string
  location_type: string
  age: number
  gender: string
  grade_level: number
  school_type: string
  distance_to_school_km: number
  mode_of_transport: string
  household_income_monthly_usd: number
  mother_education_level: string
  father_education_level: string
  family_size: number
  number_of_siblings: number
  is_orphan: boolean
  single_parent_household: boolean
  household_has_electricity: boolean
  household_has_internet_access: boolean
  owns_smartphone_or_computer: boolean
  child_labor_involvement: boolean
  attendance_rate_pct: number
  average_test_score_pct: number
  previous_year_pass: boolean
  number_of_school_transfers: number
  disciplinary_incidents_count: number
  bullying_incidents_reported: boolean
  health_issues_reported: boolean
  special_needs_status: boolean
  free_meal_program_enrolled: boolean
  scholarship_received: boolean
  extracurricular_participation: boolean
  teacher_student_ratio: number
  school_infrastructure_score: number
  early_marriage_risk_flag: boolean
  community_conflict_zone: boolean
  seasonal_migration_family: boolean
  ngo_intervention_present: boolean
  literacy_program_enrolled: boolean
  language_barrier_flag: boolean
  report_source: string
}

export interface ModelSchema {
  fields: string[]
  categorical_options: Record<string, string[]>
  numeric_bounds: Record<string, [number, number]>
  excluded_fields: string[]
  class_labels: string[]
}

export interface PredictionResponse {
  prediction: {
    risk_level: string
    probabilities: Record<string, number>
  }
  explanation?: {
    important_factors: { feature: string; importance: number }[]
    summary: string
  }
  model: {
    version: string
    selected_model: string
    primary_metric?: string
  }
}

File: datasphere/api/schemas/student.py
"""Pydantic schemas for student prediction requests."""

from __future__ import annotations

from datetime import date
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from datasphere.api.schema_fields import CATEGORICAL_OPTIONS, NUMERIC_BOUNDS, PREDICT_API_FIELDS


class StudentPredictionRequest(BaseModel):
    """Approved inference schema — no target, leakage, or identifier fields."""

    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    record_generated_date: date
    enrollment_date: date
    academic_year: str
    country: str
    state_province: str
    location_type: str
    age: int = Field(ge=5, le=20)
    gender: str
    grade_level: int = Field(ge=1, le=12)
    school_type: str
    distance_to_school_km: float = Field(ge=0, le=50)
    mode_of_transport: str
    household_income_monthly_usd: float = Field(ge=0, le=5000)
    mother_education_level: str
    father_education_level: str
    family_size: float = Field(ge=1, le=15)
    number_of_siblings: int = Field(ge=0, le=12)
    is_orphan: bool
    single_parent_household: bool
    household_has_electricity: bool
    household_has_internet_access: bool
    owns_smartphone_or_computer: bool
    child_labor_involvement: bool
    attendance_rate_pct: float = Field(ge=0, le=100)
    average_test_score_pct: float = Field(ge=0, le=100)
    previous_year_pass: bool
    number_of_school_transfers: int = Field(ge=0, le=6)
    disciplinary_incidents_count: int = Field(ge=0, le=6)
    bullying_incidents_reported: bool
    health_issues_reported: bool
    special_needs_status: bool
    free_meal_program_enrolled: bool
    scholarship_received: bool
    extracurricular_participation: bool
    teacher_student_ratio: float = Field(ge=1, le=100)
    school_infrastructure_score: float = Field(ge=0, le=100)
    early_marriage_risk_flag: bool
    community_conflict_zone: bool
    seasonal_migration_family: bool
    ngo_intervention_present: bool
    literacy_program_enrolled: bool
    language_barrier_flag: bool
    report_source: str

    @field_validator(
        "academic_year", "country", "state_province", "location_type", "gender",
        "school_type", "mode_of_transport", "mother_education_level",
        "father_education_level", "report_source",
    )
    @classmethod
    def validate_categorical(cls, value: str, info) -> str:
        allowed = CATEGORICAL_OPTIONS.get(info.field_name)
        if allowed and value not in allowed:
            raise ValueError(f"{info.field_name} must be one of: {allowed}")
        return value

    def to_inference_dict(self) -> dict:
        data = self.model_dump()
        data["record_generated_date"] = self.record_generated_date.isoformat()
        data["enrollment_date"] = self.enrollment_date.isoformat()
        return data

    @classmethod
    def field_names(cls) -> list[str]:
        return list(PREDICT_API_FIELDS)


class ModelSchemaResponse(BaseModel):
    fields: list[str]
    categorical_options: dict[str, list[str]]
    numeric_bounds: dict[str, list[float]]
    excluded_fields: list[str]
    class_labels: list[str]
